{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "import os\n",
    "os.chdir('..')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING (aesara.configdefaults): g++ not available, if using conda: `conda install m2w64-toolchain`\n",
      "WARNING (aesara.configdefaults): g++ not detected!  Aesara will be unable to compile C-implementations and will default to Python. Performance may be severely degraded. To remove this warning, set Aesara flags cxx to an empty string.\n",
      "WARNING (aesara.tensor.blas): Using NumPy C-API based implementation for BLAS functions.\n"
     ]
    }
   ],
   "source": [
    "import cloudpickle\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scipy\n",
    "import scipy.stats\n",
    "import arviz as az\n",
    "\n",
    "import pymc as pm\n",
    "import aesara\n",
    "import aesara.tensor as at\n",
    "aesara.config.exception_verbosity='high'\n",
    "\n",
    "import cobra\n",
    "import emll\n",
    "\n",
    "\n",
    "## helper functions\n",
    "def create_gprdict(model):   \n",
    "    gpr_dict = dict()\n",
    "    for rxn in model.reactions:\n",
    "        if rxn.gene_reaction_rule:\n",
    "            temp = set()\n",
    "            for x in [x.strip('() ') for x in rxn.gene_reaction_rule.split(' or ')]:\n",
    "                temp.add(frozenset(y.strip('() ') for y in x.split(' and ')))\n",
    "            gpr_dict[rxn.id] = temp\n",
    "    return gpr_dict\n",
    "\n",
    "def transcript_value_for_rxn(model, transcriptomics_df, rxn):\n",
    "    final_transcript_value = 0\n",
    "    gene_ids = []\n",
    "    for parallel_gene in create_gprdict(model)[rxn.id]:\n",
    "        transcript_values = []\n",
    "        for gene in parallel_gene:\n",
    "            if gene in transcriptomics_df.index:\n",
    "                transcript_values.append(transcriptomics_df.loc[gene])\n",
    "                # transcript_values.append(transcriptomics_df.loc[gene].to_numpy()[0])\n",
    "#                 print(transcriptomics_df.loc[gene].to_numpy()[0])\n",
    "            else:\n",
    "                transcript_values.append(np.inf)\n",
    "            min_transcript_val = np.min(transcript_values)\n",
    "        final_transcript_value = final_transcript_value + min_transcript_val\n",
    "#         if final_transcript_value==newinfbound:\n",
    "#             display(rxn.id)\n",
    "#             gene_ids.append(rxn.id)\n",
    "    return final_transcript_value\n",
    "\n",
    "def runBMCA(runName, N_ITERATIONS=1000):\n",
    "\n",
    "    name_of_script = runName\n",
    "    # import boundary-implicit, cobra-friendly version of model\n",
    "    cobra_model = cobra.io.read_sbml_model('data/iJN1463_JS.xml')\n",
    "\n",
    "    v_df = pd.read_csv(\"data/iJN1463_JS_eflux2_flux.csv\", index_col='Unnamed: 0').transpose()\n",
    "\n",
    "    # remove reactions from dataset that are not in the cobra_model\n",
    "    v_df = v_df[[i for i in v_df.columns if i in [ii.id for ii in cobra_model.reactions]]]\n",
    "\n",
    "    # import data\n",
    "    transcriptomics_df = pd.read_csv(\"data/putida_RNAseq_data.csv\", index_col='strains')[['genes', 'Value']]\n",
    "    transcriptomics_df = pd.pivot_table(transcriptomics_df, values='Value', index='genes', columns='strains').round(2)\n",
    "\n",
    "    # check if reaction has a gene reaction rule\n",
    "    geneRxns = [i.id for i in cobra_model.reactions if i.gene_reaction_rule]\n",
    "\n",
    "    # get the transcriptomics values for each reaction listed in geneRxns\n",
    "    ds = []\n",
    "    for strain in transcriptomics_df.columns:\n",
    "        # get xscript values for each reaction\n",
    "        transcriptValues = [transcript_value_for_rxn(cobra_model, transcriptomics_df[strain], cobra_model.reactions.get_by_id(i)) for i in geneRxns]\n",
    "        transcriptValues = dict(zip(geneRxns, transcriptValues))\n",
    "        # get rid of reactions with infinite transcription\n",
    "        transcriptValues = {k: v for k, v in transcriptValues.items() if v != np.inf}\n",
    "        ds.append(transcriptValues)\n",
    "    e_df = pd.DataFrame(ds, index=transcriptomics_df.columns)\n",
    "\n",
    "    # perform additive smoothing\n",
    "    ADD = 1.01\n",
    "    e_df = np.log(e_df.astype('float') + ADD)\n",
    "\n",
    "    # importing external metabolite concentration data\n",
    "    external_met_file = 'data/putida_ext_metabolomics_data.csv'\n",
    "    y_df = pd.read_csv(external_met_file)#.astype(float)\n",
    "    y_df = y_df.set_index('Line Description')\n",
    "    y_df.rename(columns={'4-Aminocinnamic acid': '4aca_e', \n",
    "                            '4-Hydroxycinnamic acid': 'pHCA_e',\n",
    "                            '4-aminobenzoic acid': '4abz_e',\n",
    "                            'CINNAMIC ACID': 'cinm_e',\n",
    "                            'p-Aminophenylalanine': '4aPhe_e'}, inplace=True)\n",
    "\n",
    "    # zero out any negative concentration values\n",
    "    y_df[y_df < 0] = 0\n",
    "\n",
    "    # Get rid of the strains in the dataset that are not present in the \n",
    "    # transcriptomics dataset.\n",
    "    y_df = y_df.loc[e_df.index]\n",
    "\n",
    "    # perform additive smoothing\n",
    "    y_df = np.log(y_df.astype('float') + ADD)\n",
    "\n",
    "    # Designate the reference strain\n",
    "    \"\"\"\n",
    "    Since we are maximizing for 4aca_e, our reference strain will be the \n",
    "    strain that produces the most 4aca_e, which is `'pACA production 3 \n",
    "    scRNA positive control'`\n",
    "    \"\"\"\n",
    "    ref_strain = 'pACA production 3 scRNA positive control'\n",
    "    v_star = v_df.loc[ref_strain].values\n",
    "    e_star = e_df.loc[ref_strain].values\n",
    "    y_star = y_df.loc[ref_strain].values\n",
    "\n",
    "    # Checking for 0 or negative values in inputs\n",
    "    assert len(v_star[v_star == 0])==0 # fluxes can be (-) but concentrations cannot\n",
    "    assert len(e_star[e_star <= 0])==0\n",
    "    assert len(y_star[y_star <= 0])==0\n",
    "\n",
    "    assert (len(e_df.values[e_df.values <= 0]) == 0)\n",
    "    assert (len(y_df.values[y_df.values <= 0]) == 0)\n",
    "\n",
    "    # Normalizing the data to the reference strain\n",
    "    yn = y_df.divide(y_star)\n",
    "    vn = v_df.divide(v_star)\n",
    "    en = e_df.divide(e_star)\n",
    "\n",
    "    assert (vn == 0).sum().sum()==0\n",
    "    assert (en <= 0).sum().sum()==0\n",
    "    assert (yn <= 0).sum().sum()==0\n",
    "\n",
    "    N = cobra.util.create_stoichiometric_matrix(cobra_model)\n",
    "\n",
    "    # Correct negative flux values at the reference state\n",
    "    N[:, v_star < 0] = -1 * N[:, v_star < 0]\n",
    "    v_star = np.abs(v_star)\n",
    "\n",
    "    assert np.isclose(np.all(np.matmul(N, v_star)), 0), \"data does not describe steady state\"\n",
    "    assert(len(np.where(N@v_star >1e-6)[0]) == 0)\n",
    "\n",
    "    # Load the Cobra version of the model\n",
    "    model = cobra.io.read_sbml_model('data/iJN1463_JS.xml') \n",
    "    model.tolerance = 1e-9\n",
    "\n",
    "    # Set up the Bayesian inference\n",
    "    # reactions and metabolite compartments\n",
    "    r_compartments = [r.compartments if 'e' not in r.compartments else 't' for r in model.reactions]\n",
    "\n",
    "    for rxn in model.exchanges:\n",
    "        r_compartments[model.reactions.index(rxn)] = 't'\n",
    "\n",
    "    m_compartments = [m.compartment for m in model.metabolites]\n",
    "    external_mets = [i for i in model.metabolites if i.compartment=='e']\n",
    "    rxnNames = [i.id for i in model.reactions]\n",
    "    v_inds = np.arange(0,len(v_star))\n",
    "    n_exp = v_df.shape[0]\n",
    "\n",
    "    # Set up elasticity matrices\n",
    "    Ex = emll.create_elasticity_matrix(model)\n",
    "    Ey = np.zeros((623,(y_df.shape[1])))\n",
    "    ey_indices = {}\n",
    "    for met in y_df.columns:\n",
    "        ey_indices[met]=[model.reactions.index(rxn) for rxn in model.metabolites.get_by_id(met).reactions]\n",
    "    for i, met in enumerate(ey_indices.keys()):\n",
    "        for ii in ey_indices[met]:\n",
    "            Ey[ii, i] = 1\n",
    "\n",
    "    # Establish labels for metabolite and reaction names\n",
    "    m_labels = [m.id for m in model.metabolites]\n",
    "    r_labels = [r.id for r in model.reactions]\n",
    "    y_labels = [i.id for i in external_mets]\n",
    "\n",
    "    ex_labels = np.array([['$\\epsilon_{' + '{0},{1}'.format(rlabel, mlabel) + '}$'\n",
    "                        for mlabel in m_labels] for rlabel in r_labels]).flatten()\n",
    "    ey_labels = np.array([['$\\epsilon_{' + '{0},{1}'.format(rlabel, mlabel) + '}$'\n",
    "                        for mlabel in y_labels] for rlabel in r_labels]).flatten()\n",
    "\n",
    "    e_labels = np.hstack((ex_labels, ey_labels))\n",
    "\n",
    "    # Setting up the PyMC model\n",
    "    ll = emll.LinLogLeastNorm(N,Ex,Ey,v_star, driver = 'gelsy')\n",
    "    from emll.util import initialize_elasticity\n",
    "\n",
    "    with pm.Model() as pymc_model:\n",
    "        # Initialize elasticities\n",
    "        Ex_t = pm.Deterministic('Ex', initialize_elasticity(N, 'ex', b=0.05, sigma=1, alpha=5, m_compartments=m_compartments,\n",
    "            r_compartments=r_compartments))\n",
    "        Ey_t = pm.Deterministic('Ey', initialize_elasticity(-Ey.T, 'ey', b=0.05, sigma=1, alpha=5))\n",
    "\n",
    "    known_e_inds = []\n",
    "    unknown_e_inds = []\n",
    "    for i, e in enumerate(rxnNames):\n",
    "        if e in e_df.columns:\n",
    "            known_e_inds.append(i)\n",
    "        else: \n",
    "            unknown_e_inds.append(i)\n",
    "    e_inds = np.hstack([known_e_inds, unknown_e_inds]).argsort()\n",
    "\n",
    "    with pymc_model:\n",
    "        #Protein Expression Priors\n",
    "        e_measured = pm.Normal('e_measured', mu=en, sigma=0.1, shape=(n_exp, len(known_e_inds))) # (41, 524)\n",
    "        e_unmeasured = pm.Laplace('e_unmeasured', mu=0, b=0.1, shape = (n_exp, len(unknown_e_inds))) # 41, 43\n",
    "\n",
    "        en_t = aesara.tensor.concatenate([e_measured, e_unmeasured], axis=1)[:, e_inds]\n",
    "        pm.Deterministic('en_t', en_t)\n",
    "\n",
    "    with pymc_model:\n",
    "        xn_t = pm.Normal('xn_t', mu=0, sigma=10, shape=(Ex.shape[1], n_exp),\n",
    "                        initval=0.1 * np.random.randn(Ex.shape[1], n_exp))\n",
    "\n",
    "    known_y_inds = []\n",
    "    omitted_y_inds = []\n",
    "    for i, y in enumerate(external_mets):\n",
    "        if y.id in y_df.columns:\n",
    "            known_y_inds.append(i)\n",
    "        else: \n",
    "            omitted_y_inds.append(i)\n",
    "    y_inds = np.hstack([known_y_inds, omitted_y_inds]).argsort()\n",
    "                    \n",
    "    # y variable should be split because there are some observations\n",
    "    with pymc_model:\n",
    "        y_unmeasured = pm.Normal('y_unmeasured', mu=3, sigma=0.1, shape = (len(omitted_y_inds), n_exp))\n",
    "\n",
    "    with pymc_model:\n",
    "        # External Metabolite and Flux Steady State priors\n",
    "        y_ss, vn_ss = ll.steady_state_aesara(Ey_t, Ex_t, en_t, xn_t.T, n_exp=n_exp)\n",
    "        pm.Deterministic('y_ss', y_ss)\n",
    "        pm.Deterministic('vn_ss', vn_ss)\n",
    "\n",
    "    with pymc_model:\n",
    "        #External Metabolite Priors\n",
    "        y_clip = y_ss[:,list(range(len(y_df.columns)))].clip(-1.5, 1.5)\n",
    "        y_obs = pm.Normal('y_measured', mu=y_clip, sigma=0.1, shape=(n_exp, yn.shape[1]), observed=yn.clip(lower=-1.5, upper=1.5)) # sigma could be y_err\n",
    "        # y_measured = pm.Normal('y_measured', mu=3, sigma=0.1, observed=yn)# shape=(n_exp, len(y_inds))\n",
    "        vn_t = pm.Normal('vn_t', mu=vn_ss, sigma=0.1, observed=vn.clip(lower=-1.5, upper=1.5))\n",
    "\n",
    "    # prior predictive check\n",
    "    with pymc_model:\n",
    "        RANDOM_SEED = np.random.seed(1)\n",
    "        trace_prior = pm.sample_prior_predictive(samples=1000, random_seed=RANDOM_SEED)\n",
    "\n",
    "    # trace_prior['prior']['ex_kinetic_entries']\n",
    "    priorEx = np.squeeze(trace_prior['prior']['Ex'].to_numpy()) # (500, 13, 8)\n",
    "\n",
    "    met_priors = np.array([ll.metabolite_control_coefficient(Ex=ex) for ex in priorEx]) \n",
    "    met_priors_hdi = az.hdi(az.convert_to_dataset(met_priors[np.newaxis,:])) #(13, 13, 2)\n",
    "\n",
    "    FCC_priors = np.array([ll.flux_control_coefficient(Ex=ex) for ex in priorEx]) # (500, 13, 13)\n",
    "    FCC_priors_hdi = az.hdi(az.convert_to_dataset(FCC_priors[np.newaxis,:])) #(13, 13, 2)\n",
    "\n",
    "    # Running the ADVI\n",
    "    \n",
    "    with pymc_model:\n",
    "        advi = pm.ADVI()\n",
    "        tracker = pm.callbacks.Tracker(\n",
    "            mean = advi.approx.mean.eval,\n",
    "            std = advi.approx.std.eval\n",
    "        )\n",
    "        approx = advi.fit(\n",
    "            n=N_ITERATIONS, \n",
    "            callbacks = [tracker],\n",
    "            obj_optimizer=pm.adagrad_window(learning_rate=5E-3), \n",
    "            total_grad_norm_constraint=0.7,\n",
    "            obj_n_mc=1)\n",
    "\n",
    "    SAMPLE_DRAWS = 1000\n",
    "\n",
    "    with pymc_model:\n",
    "        trace_vi = approx.sample(draws=SAMPLE_DRAWS, random_seed=1) \n",
    "        ppc_vi = pm.sample_posterior_predictive(trace_vi, random_seed=1)\n",
    "\n",
    "    # ppc_vi['posterior_predictive']['v_hat_obs']\n",
    "\n",
    "    ## PICKLE PICKLE PICKLE\n",
    "    cloudpickle.dump({'advi': advi,\n",
    "        'approx': approx,\n",
    "        'trace': trace_vi,\n",
    "        'trace_prior': trace_prior,\n",
    "        'e_labels': e_labels,\n",
    "        'r_labels': r_labels,\n",
    "        'm_labels': m_labels,\n",
    "        'll': ll}, file=open(f'{name_of_script}.pgz', \"wb\"))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Set parameter Username\n",
      "Academic license - for non-commercial use only - expires 2024-10-23\n"
     ]
    },
    {
     "ename": "TypeError",
     "evalue": "initialize_elasticity() got an unexpected keyword argument 'sigma'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[1;32mIn [3], line 1\u001b[0m\n\u001b[1;32m----> 1\u001b[0m \u001b[43mrunBMCA\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mtest\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mN_ITERATIONS\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m1000\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "Cell \u001b[1;32mIn [2], line 183\u001b[0m, in \u001b[0;36mrunBMCA\u001b[1;34m(runName, N_ITERATIONS)\u001b[0m\n\u001b[0;32m    179\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01memll\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mutil\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m initialize_elasticity\n\u001b[0;32m    181\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m pm\u001b[38;5;241m.\u001b[39mModel() \u001b[38;5;28;01mas\u001b[39;00m pymc_model:\n\u001b[0;32m    182\u001b[0m     \u001b[38;5;66;03m# Initialize elasticities\u001b[39;00m\n\u001b[1;32m--> 183\u001b[0m     Ex_t \u001b[38;5;241m=\u001b[39m pm\u001b[38;5;241m.\u001b[39mDeterministic(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mEx\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[43minitialize_elasticity\u001b[49m\u001b[43m(\u001b[49m\u001b[43mN\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mex\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mb\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m0.05\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43msigma\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m1\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43malpha\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;241;43m5\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mm_compartments\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mm_compartments\u001b[49m\u001b[43m,\u001b[49m\n\u001b[0;32m    184\u001b[0m \u001b[43m        \u001b[49m\u001b[43mr_compartments\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mr_compartments\u001b[49m\u001b[43m)\u001b[49m)\n\u001b[0;32m    185\u001b[0m     Ey_t \u001b[38;5;241m=\u001b[39m pm\u001b[38;5;241m.\u001b[39mDeterministic(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mEy\u001b[39m\u001b[38;5;124m'\u001b[39m, initialize_elasticity(\u001b[38;5;241m-\u001b[39mEy\u001b[38;5;241m.\u001b[39mT, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mey\u001b[39m\u001b[38;5;124m'\u001b[39m, b\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0.05\u001b[39m, sigma\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m1\u001b[39m, alpha\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m5\u001b[39m))\n\u001b[0;32m    187\u001b[0m known_e_inds \u001b[38;5;241m=\u001b[39m []\n",
      "\u001b[1;31mTypeError\u001b[0m: initialize_elasticity() got an unexpected keyword argument 'sigma'"
     ]
    }
   ],
   "source": [
    "runBMCA(\"test\", N_ITERATIONS=1000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "new_Gayles",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
